sycl/include/CL/sycl/builtins.hpp

```diff
@@ -27,95 +27,220 @@ namespace __sycl_std = __host_std;
+#define __FAST_MATH_GENFLOAT(T) (detail::is_svgenfloat<T>::value)
 #endif
 
 /* ----------------- 4.13.3 Math functions. ---------------------------------*/
+#define __SYCL_MATH_FUNCTION_OVERLOAD(NAME)                                    \
+  template <typename T, size_t N>                                              \
+  inline __SYCL_ALWAYS_INLINE                                                  \
+      std::enable_if_t<detail::is_sgenfloat<T>::value, sycl::marray<T, N>>     \
+      NAME(sycl::marray<T, N> x) __NOEXC {                                     \
+    sycl::marray<T, N> res;                                                    \
+    auto x_vec2 = reinterpret_cast<sycl::vec<T, 2> const *>(&x);               \
+    auto res_vec2 = reinterpret_cast<sycl::vec<T, 2> *>(&res);                 \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
+      res_vec2[i] = __sycl_std::__invoke_##NAME<sycl::vec<T, 2>>(x_vec2[i]);   \
+    }                                                                          \
+    if (N % 2) {                                                               \
+      res[N - 1] = __sycl_std::__invoke_##NAME<T>(x[N - 1]);                   \
+    }                                                                          \
+    return res;                                                                \
+  }
+
+__SYCL_MATH_FUNCTION_OVERLOAD(sin)
+__SYCL_MATH_FUNCTION_OVERLOAD(cos)
+__SYCL_MATH_FUNCTION_OVERLOAD(tan)
+__SYCL_MATH_FUNCTION_OVERLOAD(cospi)
+__SYCL_MATH_FUNCTION_OVERLOAD(sinpi)
+__SYCL_MATH_FUNCTION_OVERLOAD(tanpi)
+__SYCL_MATH_FUNCTION_OVERLOAD(sinh)
+__SYCL_MATH_FUNCTION_OVERLOAD(cosh)
+__SYCL_MATH_FUNCTION_OVERLOAD(tanh)
+__SYCL_MATH_FUNCTION_OVERLOAD(asin)
+__SYCL_MATH_FUNCTION_OVERLOAD(acos)
+__SYCL_MATH_FUNCTION_OVERLOAD(atan)
+__SYCL_MATH_FUNCTION_OVERLOAD(asinpi)
+__SYCL_MATH_FUNCTION_OVERLOAD(acospi)
+__SYCL_MATH_FUNCTION_OVERLOAD(atanpi)
+__SYCL_MATH_FUNCTION_OVERLOAD(asinh)
+__SYCL_MATH_FUNCTION_OVERLOAD(acosh)
+__SYCL_MATH_FUNCTION_OVERLOAD(atanh)
+__SYCL_MATH_FUNCTION_OVERLOAD(cbrt)
+__SYCL_MATH_FUNCTION_OVERLOAD(ceil)
+__SYCL_MATH_FUNCTION_OVERLOAD(floor)
+__SYCL_MATH_FUNCTION_OVERLOAD(erfc)
+__SYCL_MATH_FUNCTION_OVERLOAD(erf)
+__SYCL_MATH_FUNCTION_OVERLOAD(exp)
+__SYCL_MATH_FUNCTION_OVERLOAD(exp2)
+__SYCL_MATH_FUNCTION_OVERLOAD(exp10)
+__SYCL_MATH_FUNCTION_OVERLOAD(expm1)
+__SYCL_MATH_FUNCTION_OVERLOAD(tgamma)
+__SYCL_MATH_FUNCTION_OVERLOAD(lgamma)
+__SYCL_MATH_FUNCTION_OVERLOAD(log)
+__SYCL_MATH_FUNCTION_OVERLOAD(log2)
+__SYCL_MATH_FUNCTION_OVERLOAD(log10)
+__SYCL_MATH_FUNCTION_OVERLOAD(log1p)
+__SYCL_MATH_FUNCTION_OVERLOAD(logb)
+__SYCL_MATH_FUNCTION_OVERLOAD(rint)
+__SYCL_MATH_FUNCTION_OVERLOAD(round)
+__SYCL_MATH_FUNCTION_OVERLOAD(sqrt)
+__SYCL_MATH_FUNCTION_OVERLOAD(rsqrt)
+__SYCL_MATH_FUNCTION_OVERLOAD(trunc)
+
+#undef __SYCL_MATH_FUNCTION_OVERLOAD
+
+#define __SYCL_MATH_FUNCTION_2_OVERLOAD(NAME)                                  \
```

Comment:
Would something like this https://godbolt.org/z/adez76fTd be possible here to avoid duplicating this code for all 3 cases?

Reply:
Yes I think that something along that line would be possible, although I'm not sure that it would necessarily be an improvement, particularly for the current implementation.

Reply:
Having seen (almost?) the same `for` loop nine times in this PR, I'd argue it will be. Other reviewers, am I really the only one?

Reply:
There's a couple of issues with applying the suggested approach in the simple example https://godbolt.org/z/adez76fTd. Firstly the two `__invoke_*` function calls that are made in each `marray` math function implementation require the explicit provision of a template parameter that is different in each of the calls, i.e. `vec<T, 2>` and `T`. This means that an immediate adaption of https://godbolt.org/z/adez76fTd that directly called the `__invoke_*` functions would required the provision of two lambdas, `Callable Fvec` and `Callable F` corresponding to the `vec<T, 2>` and `T` cases.
Also with this approach the usage of macros to prevent duplicating function declaration lines becomes less attractive because this would require passing the lambdas to the macros, although I imagine that you probably meant to remove the macros completely.

I brought this issue up with the team and everyone agreed that although it would be possible to work around these issues the resultant implementation would be more complex and the size of the code would be similar to the current implementation.

I have applied your change to `to_vec2` that simplifies it.

Thanks